src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,755 @@
+		return NULL;
+	}
+
+	read = bview.len;
+	d_iov_set(&iov, bview.buf, read);
+
+	d_sg_list_t sgl = {
+	    .sg_nr     = 1,
+	    .sg_nr_out = 0,
+	    .sg_iovs   = &iov,
+	};
+
+	rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &obj, NULL, NULL);
+	if (rc) {
+		D_ERROR("Could not lookup '%s': %s (rc=%ld)", path, strerror(rc), rc);
+		rc = -rc;
+		goto out;
+	}
+
+	rc = dfs_read(hdl->dfs, obj, &sgl, 0 /* offset */, &read, NULL);
+	if (rc) {
+		goto out;
+	}
+	if (read != bview.len) {
+		rc = EIO;
+	}
+
+out:
+	PyBuffer_Release(&bview);
+
+	if (obj) {
+		int rc2 = dfs_release(obj);
+		if (rc2) {
+			D_ERROR("Could not release object '%s': %s (rc=%d)", path, strerror(rc2),
+				rc2);
+			if (rc == 0)
+				rc = rc2;
+		}
+	}
+
+	return PyLong_FromLong(rc);
+}
+
+/* describes the IO operation */
+struct io_op {
+	daos_event_t ev;
+
+	dfs_obj_t   *obj;
+	PyObject    *item;
+
+	/* Purely for debug purpose: should not be freed as it's not the owner of the data */
+	const char  *path;
+
+	daos_size_t  size;
+
+	d_iov_t      iov;
+	d_sg_list_t  sgl;
+
+	/* Buffer view that implements Python Buffer Protocol, has to be release after use */
+	Py_buffer    buf_view;
+};
+
+static int
+start_read_op(struct dfs_handle *hdl, PyObject *item, struct io_op *op)
+{
+	assert(op != NULL);
+
+	int           rc  = 0;
+	int           rc2 = 0;
+	daos_event_t *evp = &op->ev;
+
+	PyObject     *py_path = PyTuple_GetItem(item, 0);
+	PyObject     *py_buff = PyTuple_GetItem(item, 1);
+
+	if (py_path == NULL || py_buff == NULL) {
+		D_ERROR("Each tuple must contain exactly two elements: path and bytearray");
+		return EINVAL;
+	}
+
+	const char *path = PyUnicode_AsUTF8(py_path);
+	if (path == NULL) {
+		D_ERROR("First element of a tuple does not look like a path");
+		return EINVAL;
+	}
+
+	if (PyObject_GetBuffer(py_buff, &op->buf_view, PyBUF_WRITE) == -1) {
+		D_ERROR("Buffer is not writable");
+		return EINVAL;
+	}
+
+	if (!PyBuffer_IsContiguous(&op->buf_view, 'C')) {
+		D_ERROR("Buffer for '%s' is not contiguous", path);
+		rc = EINVAL;
+		goto out;
+	}
+
+	rc = daos_event_init(evp, hdl->eq, NULL);
+	if (rc) {
+		D_ERROR("Could not init event: %s (rc=%d)", d_errstr(rc), rc);
+		rc = daos_der2errno(rc);
+		goto out;
+	}
+
+	rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &op->obj, NULL, NULL);
+	if (rc) {
+		D_ERROR("Could not lookup path '%s': %s (rc=%d)", op->path, strerror(rc), rc);
+		goto out;
+	}
+
+	op->path = path;
+	op->size = op->buf_view.len;
+	d_iov_set(&op->iov, op->buf_view.buf, op->size);
+
+	op->sgl.sg_nr     = 1;
+	op->sgl.sg_nr_out = 0;
+	op->sgl.sg_iovs   = &op->iov;
+
+	rc = dfs_read(hdl->dfs, op->obj, &op->sgl, 0 /* offset */, &op->size, &op->ev);
+	if (rc) {
+		D_ERROR("Could not start async read on '%s': %s (rc=%d)", op->path, strerror(rc),
+			rc);
+		goto out;
+	}
+
+	return 0;
+
+out:
+	PyBuffer_Release(&op->buf_view);
+
+	rc2 = daos_event_fini(&op->ev);
+	if (rc2) {
+		D_ERROR("Could not finalize event: %s (rc=%d)", d_errstr(rc2), rc2);
```

Comment:
same as before, need to set rc to rc2 if rc == 0

Reply:
for this function there's `return 0` as a happy path , if the control flow is after `out` then `rc` would containt the original error code that I'd like to return, instead of the error code during the cleanup phase.

Reply:
ah ok you are right. it's probably better to rename out: as err: in those cases.
anyway not a big deal and just a nit that can be improved later

Reply:
thanks, noted for upcoming patch!